Multiple Definition Errors When Linking Both btcd and lnd Static Libraries

When attempting to link both btcd and lnd modules (built as static libraries) into the same executable, the following linking errors occur:

```
/usr/bin/ld: modules/lnd/module.a(go.o): in function `_cgo_topofstack':
(.text+0x75ee0): multiple definition of `_cgo_topofstack'; modules/btcd/module.a(go.o):(.text+0x72160): first defined here
/usr/bin/ld: modules/lnd/module.a(go.o): in function `_cgo_panic':
/usr/lib/go/src/runtime/cgo/callbacks.go:46: multiple definition of `_cgo_panic'; modules/btcd/module.a(go.o):/usr/lib/go/src/runtime/cgo/callbacks.go:46: first defined here
/usr/bin/ld: modules/lnd/module.a(go.o): in function `crosscall2':
/usr/lib/go/src/runtime/cgo/asm_amd64.s:26: multiple definition of `crosscall2'; modules/btcd/module.a(go.o):/usr/lib/go/src/runtime/cgo/asm_amd64.s:26: first defined here
```

Similar to #90

--- a/.github/workflows/workflow.yml
+++ b/.github/workflows/workflow.yml
@@ -133,7 +133,7 @@ jobs:
             asan_options: "detect_leaks=0:detect_container_overflow=0"
           - corpus: "verify_script"
             target: "verify_script"
-            cxxflags: "-DBITCOIN_CORE -DBTCD"
+            cxxflags: "-DBITCOIN_CORE -DBTCD -DGOCOIN"
             asan_options: "detect_leaks=0:detect_container_overflow=0"
           - corpus: "descriptor_parse_clean"
             target: "descriptor_parse"

--- a/modules/btcd/Makefile
+++ b/modules/btcd/Makefile
@@ -1,5 +1,7 @@
 CXXFLAGS += -Wall -Wextra -std=c++20 -I. -I ../../include -I./btcd_wrapper
 
+SYMBOL_PREFIX ?= btcd
+
 all: module.a
 
 module.a: module.o btcd_wrapper/libbtcd_wrapper.a
@@ -8,6 +10,7 @@ module.a: module.o btcd_wrapper/libbtcd_wrapper.a
 
 btcd_wrapper/libbtcd_wrapper.a: btcd_wrapper/wrapper.go
 	cd btcd_wrapper && go build -o libbtcd_wrapper.a -buildmode=c-archive -tags=libfuzzer -gcflags=all=-d=libfuzzer wrapper.go
+	bash ../../rename_cgo_symbols.sh btcd_wrapper/libbtcd_wrapper.a $(SYMBOL_PREFIX)
 
 module.o: module.cpp module.h
 	$(CXX) $(CXXFLAGS) -fPIC -c module.cpp -o module.o

--- a/modules/decredsecp256k1/Makefile
+++ b/modules/decredsecp256k1/Makefile
@@ -1,5 +1,7 @@
 CXXFLAGS += -Wall -Wextra -std=c++20 -I. -I ../../include -I./decredsecp256k1_wrapper
 
+SYMBOL_PREFIX ?= decredsecp256k1
+
 all: module.a
 
 module.a: module.o decredsecp256k1_wrapper/libdecredsecp256k1_wrapper.a
@@ -8,6 +10,7 @@ module.a: module.o decredsecp256k1_wrapper/libdecredsecp256k1_wrapper.a
 
 decredsecp256k1_wrapper/libdecredsecp256k1_wrapper.a:
 	cd decredsecp256k1_wrapper && go build -o libdecredsecp256k1_wrapper.a -buildmode=c-archive -tags=libfuzzer -gcflags=all=-d=libfuzzer wrapper.go
+	bash ../../rename_cgo_symbols.sh decredsecp256k1_wrapper/libdecredsecp256k1_wrapper.a $(SYMBOL_PREFIX)
 
 module.o: module.cpp
 	$(CXX) $(CXXFLAGS) -fPIC -c module.cpp -o module.o

--- a/modules/gocoin/Makefile
+++ b/modules/gocoin/Makefile
@@ -1,5 +1,7 @@
 CXXFLAGS += -Wall -Wextra -std=c++20 -I. -I ../../include -I./gocoin_wrapper
 
+SYMBOL_PREFIX ?= gocoin
+
 all: module.a
 
 module.a: module.o gocoin_wrapper/libgocoin_wrapper.a
@@ -8,6 +10,7 @@ module.a: module.o gocoin_wrapper/libgocoin_wrapper.a
 
 gocoin_wrapper/libgocoin_wrapper.a:
 	cd gocoin_wrapper && go build -o libgocoin_wrapper.a -buildmode=c-archive -tags=libfuzzer -gcflags=all=-d=libfuzzer wrapper.go
+	bash ../../rename_cgo_symbols.sh gocoin_wrapper/libgocoin_wrapper.a $(SYMBOL_PREFIX)
 
 module.o: module.cpp gocoin_wrapper/libgocoin_wrapper.a
 	$(CXX) $(CXXFLAGS) -fPIC -c module.cpp -o module.o

--- a/modules/lnd/Makefile
+++ b/modules/lnd/Makefile
@@ -1,5 +1,7 @@
 CXXFLAGS += -Wall -Wextra -std=c++20 -I. -I ../../include -I./lnd_wrapper
 
+SYMBOL_PREFIX ?= lnd
+
 all: module.a
 
 module.a: module.o lnd_wrapper/liblnd_wrapper.a
@@ -8,6 +10,7 @@ module.a: module.o lnd_wrapper/liblnd_wrapper.a
 
 lnd_wrapper/liblnd_wrapper.a:
 	cd lnd_wrapper && go build -o liblnd_wrapper.a -buildmode=c-archive -tags=libfuzzer -gcflags=all=-d=libfuzzer wrapper.go
+	bash ../../rename_cgo_symbols.sh lnd_wrapper/liblnd_wrapper.a $(SYMBOL_PREFIX)
 
 golang: lnd_wrapper/liblnd_wrapper.a
 

--- a/rename_cgo_symbols.sh
+++ b/rename_cgo_symbols.sh
@@ -1,0 +1,93 @@
+#!/bin/bash
+# rename_cgo_symbols.sh - Rename CGO runtime symbols in a Go c-archive to avoid
+# duplicate symbol conflicts when linking multiple Go modules together.
+#
+# Usage: rename_cgo_symbols.sh <archive> <prefix>
+#
+# When Go builds a c-archive (-buildmode=c-archive), it bundles CGO runtime
+# support symbols (e.g. _cgo_topofstack, _cgo_panic, crosscall2) into the
+# archive. Linking two or more such archives in the same binary causes
+# "multiple definition" linker errors.
+#
+# This script renames every globally-visible CGO runtime symbol to
+# PREFIX_<original_name> using objcopy --redefine-syms. Both the definition
+# and all internal references are updated, so each module's CGO runtime
+# becomes self-contained under its own namespace.
+#
+# Module-specific CGO symbols (those containing a 12-char hex hash such as
+# _cgo_0ada0d83d011_Cfunc_free) are left untouched.
+#
+# Requires: nm, ar, ranlib, objcopy (or llvm-objcopy)
+
+set -e
+
+ARCHIVE="$1"
+PREFIX="$2"
+
+if [ -z "$ARCHIVE" ] || [ -z "$PREFIX" ]; then
+    echo "Usage: $0 <archive> <prefix>" >&2
+    exit 1
+fi
+
+if [ ! -f "$ARCHIVE" ]; then
+    echo "Error: Archive '$ARCHIVE' not found" >&2
+    exit 1
+fi
+
+# Resolve to absolute path before changing directories
+ARCHIVE="$(cd "$(dirname "$ARCHIVE")" && pwd)/$(basename "$ARCHIVE")"
+
+# Find objcopy (prefer llvm-objcopy on macOS where GNU objcopy is unavailable)
+OBJCOPY=$(command -v objcopy 2>/dev/null || command -v llvm-objcopy 2>/dev/null || echo "")
+if [ -z "$OBJCOPY" ]; then
+    echo "Error: objcopy not found. Install with: apt-get install binutils (Linux) or brew install llvm (macOS)" >&2
+    exit 1
+fi
+
+WORK_DIR=$(mktemp -d)
+trap 'rm -rf "$WORK_DIR"' EXIT
+
+MAP_FILE="$WORK_DIR/cgo_runtime.map"
+
+# Find all globally-visible CGO runtime symbols in the archive.
+#
+# CGO runtime symbols match: _cgo_*, x_cgo_*, crosscall1, crosscall2, and
+# derived names like _x_crosscall2_ptr.
+#
+# Module-specific CGO symbols contain a 12-char lowercase hex hash (e.g.
+# _cgo_0ada0d83d011_Cfunc_free or _cgoexp_0ada0d83d011_MyFunc) and must be
+# excluded so the exported Go functions remain accessible under their
+# original names.
+#
+# nm type letters: T=text D=data B=bss R=rodata S=other-section (all global)
+nm "$ARCHIVE" 2>/dev/null \
+    | awk '/ [TDBRS] / { print $NF }' \
+    | grep -E '(_cgo_|x_cgo_|crosscall)' \
+    | grep -vE '[0-9a-f]{12}' \
+    | sort -u \
+    | while IFS= read -r sym; do
+        printf '%s %s_%s\n' "$sym" "$PREFIX" "$sym"
+    done > "$MAP_FILE"
+
+if [ ! -s "$MAP_FILE" ]; then
+    echo "No CGO runtime symbols found in $(basename "$ARCHIVE"), nothing to rename."
+    exit 0
+fi
+
+COUNT=$(wc -l < "$MAP_FILE" | tr -d ' ')
+echo "Renaming $COUNT CGO runtime symbols with prefix '${PREFIX}_' in $(basename "$ARCHIVE") ..."
+
+# Extract all objects, apply the symbol rename map, repack the archive.
+OBJS_DIR="$WORK_DIR/objs"
+mkdir -p "$OBJS_DIR"
+(cd "$OBJS_DIR" && ar x "$ARCHIVE")
+
+for obj in "$OBJS_DIR"/*.o; do
+    [ -f "$obj" ] || continue
+    "$OBJCOPY" --redefine-syms="$MAP_FILE" "$obj" 2>/dev/null || true
+done
+
+ar rcs "$ARCHIVE" "$OBJS_DIR"/*.o
+ranlib "$ARCHIVE"
+
+echo "Done. CGO runtime symbols in $(basename "$ARCHIVE") now prefixed with '${PREFIX}_'."

